util/topgen.py

```diff
@@ -776,30 +776,38 @@ def render_template(template_path: str, rendered_path: Path, **other_info):
         with rendered_path.open(mode="w", encoding="UTF-8") as fout:
             fout.write(template_contents)
 
-    # The Rust file needs some complex information, so we initialize this
-    # object to store it.
-    rs_helper = TopGenRust(completecfg, name_to_block, version_stamp)
+    for addr_space in completecfg['addr_spaces']:
+        default_addr_space = addr_space.get('default', False)
 
-    rust_files = [("toplevel_mod.rs.tpl", "mod.rs"),
-                  ("toplevel.rs.tpl", f"top_{topname}.rs"),
-                  ("toplevel_memory.rs.tpl", f"top_{topname}_memory.rs")]
+        addr_space_suffix = ""
+        if not default_addr_space:
+            addr_space_suffix = "_" + addr_space['name']
 
-    # Creating Rust output directory
-    rsformat_dir = out_path / "sw/autogen/chip/"
-    rsformat_dir.mkdir(parents=True, exist_ok=True)
+        # The Rust file needs some complex information, so we initialize this
+        # object to store it.
+        rs_helper = TopGenRust(completecfg, name_to_block, version_stamp, addr_space['name'])
```

Comment:
I know that the rest of this file has the same problem, but maybe it makes sense to do some sanity checking of `addr_space['name']` here? (Make sure it's a string, maybe? And how does `default_addr_space` relate to the allowed values here? (Or maybe that sort of check belongs somewhere earlier, rather than being Rust-specific?)

Reply:
I agree, it might be checked that `addr_space['name']` is a string. There is no custom validation for these kind of dicts, right? `default_addr_space` is determined on demand from the list of address spaces, the one that specifies `default = true`.

Reply:
I don't think there is (not code I wrote...). And good point about `default_addr_space`: I totally misread your code!